### PR TITLE
[ipatests] Add basic support for remote ipaplatform

### PR DESCRIPTION
**Pull request by stanislavlevin**
The main goal is replacing the usage of controller's ipaplatform with host's one.

- added basic support (JSON-serializable attributes)  for remote platform's paths, knownservices, constants, osinfo. Expose them as `host.ipaplatform.ATTR`
- added wrapper for remote systemctl which autoresolves knownservices of IPA
- added convenient host markers `skip_if_hostplatform`, `skip_if_hostcontainer`, `skip_if_hostfips`, ` skip_if_not_hostselinux`, `skip_if_host`. Host markers being applied at class scope skip the installation phase.

Fixes: https://pagure.io/freeipa/issue/8620

### CODE REVIEW

File: ipatests/pytest_ipa/integration/host_ipaplatform.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+
+                    remote_osinfo = {}
+                    for name in sorted(dir(osinfo)):
+                        if name.startswith("_"):
+                            continue
+
+                        value = getattr(osinfo, name)
+                        try:
+                            json.dumps(value)
```
Comment (flo-renaud) [inline ipatests/pytest_ipa/integration/host_ipaplatform.py:171]: For the version_number attribute (which is a tuple), this call to json.dumps + json.loads transforms the tuple into a list:
```
>>> import json
>>> from ipaplatform.osinfo import osinfo
>>> value = getattr(osinfo, 'version_number')
>>> value
(33,)
>>> json.loads(json.dumps(value))
[33]
```
and that later creates an issue in the test:
```    def test_smb_service_s4u2self(self):
        """Test S4U2Self operation by IPA service
           against both AD and IPA users
        """
        master_os = self.master.ipaplatform.osinfo
>       if master_os.id == "fedora" and master_os.version_number <= (31,):
E       TypeError: '<=' not supported between instances of 'list' and 'tuple'

test_integration/test_smb.py:338: TypeError
```

**Review by flo-renaud — COMMENTED** [review]
Hi @stanislavlevin 
I re-launched the 3 tests that failed on timeout/vagrant issues (`simple_replication`, `test_caless_TestReplicaInstall` and `test_trust`)

About the other failures:
- `test_installation_TestInstallWithCA_DNS3` is a known issue: [8700](https://pagure.io/freeipa/issue/8700)
- `test_ntp_options`: [8676](https://pagure.io/freeipa/issue/8676)
- `test_replica_promotion_TestHiddenReplicaPromotion`: [8582](https://pagure.io/freeipa/issue/8582)
- `test_topology_TestCASpecificRUVs`: [8023](https://pagure.io/freeipa/issue/8023)
- the other are new failures

Please find inline comments that may help fix some of the issues

**Comment by stanislavlevin** [comment]
@flo-renaud, hi!
Thanks a lot for looking at this, tbh I didn't check the failures yet and just triggered the nightly to gather all unfixed issues.

There are many boring changes and I understand that the review is problematic. If there are no strong objections I would proceed. I planned to cover some functionality with tests.

Of course, any comments would be very appreciated.
Thank you!

File: ipatests/test_integration/test_caless.py [hunk 2]
```diff
@@ -76,16 +75,40 @@ def get_replica_prepare_stdin(cert_passwords=()):
...
+    )
     # A workaround for https://fedorahosted.org/freeipa/ticket/4639
-    result = host.run_command(['certutil', '-L', '-d',
-                               paths.HTTPD_ALIAS_DIR], raiseonerr=False)
+    result = host.run_command(
+        [
+            "certutil",
+            "-L",
+            "-d",
+            host.ipalatform.paths.HTTPD_ALIAS_DIR,
```
[the remark this answers is not in the document]
Comment (rcritten) [reply ipatests/test_integration/test_caless.py:95]: I'd argue this code can be removed completely. We haven't supported using mod_nss for years now.

**Comment by abbra** [comment]
I think this is a very important change to help with cross-distribution tests, so your effort on this is very welcome.

Please rebase as there are conflicts on few test files that were modified meanwhile.

**Comment by stanislavlevin** [comment]
Working on mypy adaptation.

**Comment by stanislavlevin** [comment]
There is magic "/usr/sbin/sssd: error while loading shared libraries: /lib64/libtdb.so.1: invalid ELF header" on failed machines.

**Comment by stanislavlevin** [comment]
Rebased.

**Comment by stanislavlevin** [comment]
I suspect that members don't know/like mypy. So, I can move this part out to its own PR, right now it helps me here.

**Comment by stanislavlevin** [comment]
weekly rebase.

**Comment by stanislavlevin** [comment]
It's not stale.
Waiting for any comment.
It takes a long time to rebase it every time and recheck again.
I'm ok with dropping mypy part since nobody wants it.

**Comment by abbra** [comment]
Can you please rebase to master branch?

**Comment by stanislavlevin** [comment]
I'm a bit overloaded :( will try to take this next week.

**Comment by abbra** [comment]
@stanislavlevin sure, first couple weeks on January would be OK too.

**Comment by stanislavlevin** [comment]
`test_server_del` failures are tracked there:
https://pagure.io/freeipa/issue/8865

**Comment by stanislavlevin** [comment]
rebased to today's master.

**Comment by stanislavlevin** [comment]
rebased

**Comment by stanislavlevin** [comment]
rebased

**Comment by f-trivino** [comment]
> rebased

@stanislavlevin thanks for the rebase :)

I'm doing some manual testing in addition, involving rhel/centos, to exercise a bit different remote hosts than fedora. In the mean time, I triggered some additional tests here: https://github.com/freeipa-pr-ci2/freeipa/pull/1525

I found a couple of issues:

fedora-latest/test_cert_fix — <no description>
[Details](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/042e45c0-9427-11ec-bb1a-fa163ecec9cc)
@freeipa-pr-ci2
fedora-latest/test_ipahealthcheck — <no description>
[Details](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/f94fb01c-9426-11ec-aeb7-fa163e7c9f91)

but they are probably related to the missing rebase. Let's see if they passes now.

**Comment by stanislavlevin** [comment]
rebased.

**Comment by abbra** [comment]
@stanislavlevin could you please rebase? There are few conflicts now.

**Comment by stanislavlevin** [comment]
Working on it. I need to adapt the code to changes made in the last half of this year.

**Comment by flo-renaud** [comment]
Some of the failing tests are known issues, not related to this PR:
- `test_fips`: [9002](https://pagure.io/freeipa/issue/9002)
- `test_installation_TestInstallWithCA_DNS3`: [9135](https://pagure.io/freeipa/issue/9135)
- `test_smb`: [9124](https://pagure.io/freeipa/issue/9124)
- `test_topology_TestCASpecificRUVs`: [8023](https://pagure.io/freeipa/issue/8023)
- `test_webui_policy`: [9161](https://pagure.io/freeipa/issue/9161)
- `test_webui_subid`: [9214](https://pagure.io/freeipa/issue/9214)
- `test_webui_users`: [9212](https://pagure.io/freeipa/issue/9212)

The other failures are probably infra issues, I relaunched the tests. Let's wait and see the results on 2nd run.

**Pull request closed without merging**